cow_instrument/ComponentInfo.h

```diff
@@ -47,82 +39,181 @@ template <typename InstTree> class ComponentInfo {
+    for (auto &index : componentIndexes) {
+      pos += this->position(index);
+    }
+    pos /= nextLevelSize;
+    return pos;
+  } else {
+    // Is either a path component or a detector
+    auto detIndex = m_componentToDetectorIndex[componentIndex];
+    if (detIndex >= 0) {
+      return m_detectorInfo->position(detIndex);
+    } else {
+      auto pathIndex = m_componentToPathIndex[componentIndex];
+
+      return m_detectorInfo->pathComponentInfo().position(pathIndex);
+    }
+  }
 }
 
 template <typename InstTree>
 Eigen::Quaterniond
 ComponentInfo<InstTree>::rotation(size_t componentIndex) const {
-  return (*m_rotations)[componentIndex];
+  const std::vector<size_t> componentIndexes =
+      m_instrumentTree.nextLevelIndexes(componentIndex);
+  const auto nextLevelSize = componentIndexes.size();
+  if (nextLevelSize > 0) {
+    // We are dealing with a composite
+    throw std::runtime_error("Rotations for composites not implemented");
+  } else {
+    // Is either a path component or a detector
+    auto detIndex = m_componentToDetectorIndex[componentIndex];
+    if (detIndex >= 0) {
+      return m_detectorInfo->rotation(detIndex);
+    } else {
+      auto pathIndex = m_componentToPathIndex[componentIndex];
+
+      return m_detectorInfo->pathComponentInfo().rotation(pathIndex);
+    }
+  }
 }
 
+
+
 template <typename InstTree>
 size_t ComponentInfo<InstTree>::componentSize() const {
-  return m_instrumentTree->componentSize();
+  return m_instrumentTree.componentSize();
 }
 
 template <typename InstTree>
 const InstTree &ComponentInfo<InstTree>::const_instrumentTree() const {
-  return *m_instrumentTree;
+  return m_instrumentTree;
 }
 
 template <typename InstTree>
 void ComponentInfo<InstTree>::move(size_t componentIndex,
                                    const Eigen::Vector3d &offset) {
 
-  const std::vector<size_t> indexes =
-      m_instrumentTree->subTreeIndexes(componentIndex);
-  for (auto &index : indexes) {
-    (*m_positions)[index] += offset;
-    // TODO m_entry and m_exit points should also be translated here!
+  // All "connected" component indexes
+  const std::vector<size_t> componentIndexes =
+      m_instrumentTree.subTreeIndexes(componentIndex);
+
+  std::vector<size_t>
+      detectorsToMove; // Cache of detector indexes which will be moved.
+  std::vector<size_t> pathComponentsToMove; // Cache of path component indexes
+                                            // which will be moved
+
+  detectorsToMove.reserve(componentIndexes.size());
+  pathComponentsToMove.reserve(componentIndexes.size());
+  for (auto &compIndex : componentIndexes) {
+    auto detIndex = m_componentToDetectorIndex[compIndex];
+    if (detIndex >= 0) {
+      detectorsToMove.push_back(detIndex);
+    }
+
+    auto pathIndex = m_componentToPathIndex[compIndex];
+    if (pathIndex >= 0) {
+      pathComponentsToMove.push_back(pathIndex);
+    }
+
+    // Anything else in the subtree is ignored. Correctly.
```

Comment:
@SimonHeybrock That's definitely a possibility worth exploring I'll give that a go. Just to be clear - we would still need to work with the fact that we do want to store the absolute positions of those detectors in the bank, so any update to the banks position will still result in us also modifying the detector positions.